fix: duplicated endpoint per hosted zone

leonardocaylent opened the pull request:
**Description**

Looking into Issue #4241 we found that Delete was trying to generate 2 calls with 6 records per hosted zone instead of 2 calls with 3 records each. Taking a deeper look, changes.Delete at plan.go was handling duplicates at 0.13.6 but on the latest version it's not filtering them anymore.
Removing the duplicate endpoints in plan.go solves this issue for all providers

**Short explain of new behavior**
Added a new function RemoveDuplicates to detect duplicated endpoints when needed
Added Unit Tests for the function
Removed duplicated records on plan.go for changes.Delete
 

**Details**
Create behavior:
Without the fix:

```
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=debug msg="Adding cname-testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding cname-testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=info msg="Desired change: CREATE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ2]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="3 record(s) in zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2] were successfully updated"
level=info msg="Desired change: CREATE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ1]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="3 record(s) in zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1] were successfully updated"
```

With the fix:

```
level=info msg="Desired change: CREATE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ2]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="3 record(s) in zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2] were successfully updated"
level=info msg="Desired change: CREATE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ1]"
level=info msg="Desired change: CREATE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
```

Delete behavior:
Without the fix:
```
level=debug msg="Considering zone: /hostedzone/HZ1 (domain: sandbox.yourdomain.com.)"
level=debug msg="Considering zone: /hostedzone/HZ2 (domain: internal.sandbox.yourdomain.com.)"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=debug msg="Adding cname-testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding cname-testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=debug msg="Adding cname-testdeploy.internal.sandbox.yourdomain.com. to zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2]"
level=debug msg="Adding cname-testdeploy.internal.sandbox.yourdomain.com. to zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="Desired change: DELETE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ2]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ2]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=error msg="Failure in zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2] when submitting change batch: InvalidChangeBatch: [The request contains an invalid set of changes for a resource record set 'TXT cname-testdeploy.internal.sandbox.yourdomain.com.', The request contains an invalid set of changes for a resource record set 'A testdeploy.internal.sandbox.yourdomain.com.', The request contains an invalid set of changes for a resource record set 'TXT testdeploy.internal.sandbox.yourdomain.com.']\n\tstatus code: 400, request id: X"
level=info msg="Desired change: DELETE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=error msg="Failure in zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1] when submitting change batch: InvalidChangeBatch: [The request contains an invalid set of changes for a resource record set 'TXT cname-testdeploy.internal.sandbox.yourdomain.com.', The request contains an invalid set of changes for a resource record set 'A testdeploy.internal.sandbox.yourdomain.com.', The request contains an invalid set of changes for a resource record set 'TXT testdeploy.internal.sandbox.yourdomain.com.']\n\tstatus code: 400, request id: X"
level=error msg="Failed to do run once: soft error\nfailed to submit all changes for the following zones: [/hostedzone/HZ2 /hostedzone/HZ1]"
```

With the fix:
```
level=debug msg="Considering zone: /hostedzone/HZ2 (domain: internal.sandbox.yourdomain.com.)"
level=debug msg="Considering zone: /hostedzone/HZ1 (domain: sandbox.yourdomain.com.)"
level=info msg="Desired change: DELETE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ1]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ1]"
level=info msg="3 record(s) in zone sandbox.yourdomain.com. [Id: /hostedzone/HZ1] were successfully updated"
level=info msg="Desired change: DELETE cname-testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com A [Id: /hostedzone/HZ2]"
level=info msg="Desired change: DELETE testdeploy.internal.sandbox.yourdomain.com TXT [Id: /hostedzone/HZ2]"
level=info msg="3 record(s) in zone internal.sandbox.yourdomain.com. [Id: /hostedzone/HZ2] were successfully updated"
```

Fixes #4241 

**Checklist**

- [x] Unit tests updated
- [ ] End user documentation updated

mloiseleur commented:
/ok-to-test

mloiseleur commented:
/retitle fix: duplicated endpoint per hosted zone

mloiseleur commented:
@leonardocaylent Do you think you can add a test on this ? 

In order to keep issues fixed for good, we accept PR only when there is a test with it.

leonardocaylent commented:
> @leonardocaylent Do you think you can add a test on this ?
> 
> In order to keep issues fixed for good, we accept PR only when there is a test with it.

Yes, I can. I also need to check for TestAWSRecords and TestAWSApplyChanges that are failing.

yurrriq commented on provider/aws/aws.go:
```diff
@@ -993,6 +994,18 @@ func changesByZone(zones map[string]*route53.HostedZone, changeSet Route53Change
+                visitedHostnames[aws.StringValue(z.Id)] = make(map[string]bool)
+            }
+
+			if visitedHostnames[aws.StringValue(z.Id)][key] {
```
```suggestion
			if _, ok := visitedHostnames[aws.StringValue(z.Id)][key]; ok {
```

yurrriq commented on provider/aws/aws.go:
```diff
@@ -1008,7 +1021,8 @@ func changesByZone(zones map[string]*route53.HostedZone, changeSet Route53Change
 			}
 			changes[aws.StringValue(z.Id)] = append(changes[aws.StringValue(z.Id)], c)
-			log.Debugf("Adding %s to zone %s [Id: %s]", hostname, aws.StringValue(z.Name), aws.StringValue(z.Id))
+			visitedHostnames[aws.StringValue(z.Id)][key] = true
```
```suggestion
			visitedHostnames[aws.StringValue(z.Id)][key] = struct{}{}
```

yurrriq reviewed (changes requested):
We can optimize this a bit by using the empty struct.

mloiseleur commented:
Go modules update are done in dedicated PRs. 

@leonardocaylent Would you please remove update of go.mod & go.sum in this PR ?

leonardocaylent commented:
@mloiseleur I need to test this image and review again the visitedChanges validation for aws.go and aws_test.go. But I'd love to have some review/thoughts on this. @yurrriq thanks for your comments, I didn't apply your suggestions yet because I'm focusing on the fix approach rather than optimizations. I'll optimize it when we confirm the path we want to follow to fix this. This seemed like an S but it's an L / XL since the impact that could have. Also the coverage of endpoint.go may need to be increased

Raffo commented:
A change to endpoints.go is impacting every provider, but AFAIK we only have reports of this being a problem for a specific case for the aws provider. Is it impacting generally every provider?

Raffo commented on endpoint/endpoint.go:
```diff
@@ -293,12 +293,20 @@ func (e *Endpoint) String() string {
 	filtered := []*Endpoint{}
+	visited := make(map[EndpointKey]bool) // Initialize the visited map
 	for _, ep := range eps {
+		key := EndpointKey{DNSName: ep.DNSName, RecordType: ep.RecordType, SetIdentifier: ep.SetIdentifier}
```
What are the possible cases in which we have duplicated endpoints in this loop? From my understanding, it's only for overlapping zones, is this right?

Raffo commented on provider/aws/aws.go:
```diff
@@ -993,6 +995,28 @@ func changesByZone(zones map[string]*route53.HostedZone, changeSet Route53Change
 		}
 		for _, z := range zones {
+			// IMPORTANT EXPLAIN: I tried to fix this in here but a lot of tests started to fail in cascade.
+			//                    Found that nil recordTypes are entering in this function
```
When are "nil recordTypes" entering this loop?

leonardocaylent replied:
> When are "nil recordTypes" entering this loop?

func TestAWSChangesByZones(t *testing.T) { in aws_test.go is running tests providing empty RecordTypes on the function

leonardocaylent commented:
> A change to endpoints.go is impacting every provider, but AFAIK we only have reports of this being a problem for a specific case for the aws provider. Is it impacting generally every provider?

We don't know the behavior of different providers. They should be having the same issue when applying FilterEndpointsByOwnerID

leonardocaylent commented:
@Raffo Apart from the comments I made, I found that if we fix this on endpoint.go using :

```
for _, ep := range eps {
		key := ep.Key()    <----------- I tested the function that provides unique endpoints based on RecordType Name and SetIdentifier
		// Skip duplicated endpoints
		if visited[key] {
			log.Debugf(`Skipping duplicated endpoint %v `, ep)
			continue
		}
		if endpointOwner, ok := ep.Labels[OwnerLabelKey]; !ok || endpointOwner != ownerID {
			log.Debugf(`Skipping endpoint %v because owner id does not match, found: "%s", required: "%s"`, ep, endpointOwner, ownerID)
		} else {
			filtered = append(filtered, ep)
			log.Debugf(`Added endpoint %v because owner id matches, found: "%s", required: "%s"`, ep, endpointOwner, ownerID)
		}
		if(key != EndpointKey{}){
			visited[key] = true
		}
	}
```
	
One test in service_test.go fails:
`annotated services return an endpoint with hostname then resolve hostname`
The result of the test is expected 2 returned 0
If we fix it on aws.go we need to add more conditions to handle weighted and geolocation records (this was captured by make test while I was testing the fix)
I think the next step is to think where is it worth to fix this.

leonardocaylent commented:
@Raffo I modified the validation:
I just made changes on endpoint.go and added more visibility for debug logs on aws.go

About this lines:
```
		// Adding validation (skipping empty EndpointKeys)
		// Not sure why we have cases when there are empty keys, but we do
		if key != (EndpointKey{}) {
			visited[key] = true
		}
```
Without the if statement this error throws up about the new 2 unit tests on endpoint_test.go:
```
--- FAIL: TestDuplicatedFilterEndpointsByOwnerID (0.00s)
    --- FAIL: TestDuplicatedFilterEndpointsByOwnerID/filter_values (0.00s)
        endpoint_test.go:287: FilterEndpointsByOwnerID() = [foo.com 0 IN A   [] foo.com 0 IN CNAME   [] foo.com 0 IN A   [] foo.com 0 IN CNAME   []], want [foo.com 0 IN A   [] foo.com 0 IN CNAME   []]
--- FAIL: TestZonesDuplicatedFilterEndpointsByOwnerID (0.00s)
    --- FAIL: TestZonesDuplicatedFilterEndpointsByOwnerID/filter_values (0.00s)
        endpoint_test.go:402: FilterEndpointsByOwnerID() = [internal.foo.com 0 IN A   [] internal.foo.com 0 IN CNAME   [] internal.foo.com 0 IN A   [] internal.foo.com 0 IN CNAME   [] foo.com 0 IN A   [] foo.com 0 IN CNAME   [] foo.com 0 IN A   [] foo.com 0 IN CNAME   []], want [internal.foo.com 0 IN A   [] internal.foo.com 0 IN CNAME   [] foo.com 0 IN A   [] foo.com 0 IN CNAME   []]
time="2024-03-16T18:22:57Z" level=error msg="Invalid target net filter: "
time="2024-03-16T18:22:57Z" level=error msg="Invalid target net filter: "
time="2024-03-16T18:22:57Z" level=error msg="Invalid target net filter: 0"
time="2024-03-16T18:22:57Z" level=error msg="Invalid target net filter: "
time="2024-03-16T18:22:57Z" level=error msg="Invalid target net filter: "
FAIL
coverage: 61.0% of statements
FAIL    sigs.k8s.io/external-dns/endpoint       0.109s
```

Even the unit-tests are passing and the behavior of overlapping zones is corrected with this code, this probably needs some extra eyes to verify this is the way we want to go or improve the code (maybe using empty structs instead of boolean for memory as @yurrriq mentioned), I'm tagging @cronik @johngmyers @szuecs @mloiseleur.

In case we don't find a solution I'd suggest to rollback the changes made on https://github.com/kubernetes-sigs/external-dns/pull/3747 and fix this later to unblock all the users of 0.14.0 that uses overlapping zones on AWS.

szuecs commented on endpoint/endpoint_test.go:
```diff
@@ -216,3 +216,191 @@ func TestIsOwnedBy(t *testing.T) {
+			},
+			want: []*Endpoint{
+				foo1,
+				foo2,
```
This result looks wrong. You can not have CNAME and A record with the same name.
Another problem is that if you have multiple IPs in your ingress status, you will want to set multiple A records for the same name.

leonardocaylent replied:
@szuecs Thank you for your review. I was taking as example the unit test on https://github.com/kubernetes-sigs/external-dns/blob/master/endpoint/endpoint_test.go#L120-L174. About the ingress status with multiple IPs issue, does that mean that we don't have unit tests for that use case and that's why I got a pass on the tests?

leonardocaylent commented:
> I suggest to remove all the changes in `aws.go`. They don't seem to be necessary. Otherwise, it looks very good to me. Thanks a lot !

That was just an addition to add more visibility on debug mode to see the record types that are being created. I'll remove it. Thanks for your feedback

leonardocaylent commented:
Based on this [comment](https://github.com/kubernetes-sigs/external-dns/pull/4296#discussion_r1532670169), this [lines](https://github.com/kubernetes-sigs/external-dns/blob/master/endpoint/endpoint_test.go#L120-L174) on master are a wrong unit test, we can't have CNAME and A records with the same name. 

I analyzed if the Pull Request that introduced this issue can be easily reverted and is not so easy, some pull requests were added above the initial [PR](https://github.com/kubernetes-sigs/external-dns/pull/3747)

So the 3 options that we have are:
1)Fix on endpoint.go & fix unit tests
2)Fix on aws.go and contemplate emptyrecords, weighted and geolocation & fix unit tests
3)Revert changes on 3747, remove the wrong unit tests and reopen the multi-cluster dual stack record types issue

Raffo commented:
I would prefer to fix forward and not revert the PR. The order I see is:

- Fix the wrong unit tests 
- fix the code based on the tests
- implement whatever fix we agree on.

the first two points are the same PR, the third a different one (possibly this one).

cronik commented:
> Based on this https://github.com/kubernetes-sigs/external-dns/pull/4296#discussion_r1532670169, this [lines](https://github.com/kubernetes-sigs/external-dns/blob/master/endpoint/endpoint_test.go#L120-L174) on master are a wrong unit test, we can't have CNAME and A records with the same name.

@leonardocaylent to be clear the referenced lines may not be technically correct according to DNS rules, but for the specific test `TestFilterEndpointsByOwnerID` it doesn't matter because the test is for the "OwnerID" filter which is not responsible detecting invalid/duplicate names. The logic to remove duplicates should probably not be added to that function.

leonardocaylent commented:
> > Based on this [#4296 (comment)](https://github.com/kubernetes-sigs/external-dns/pull/4296#discussion_r1532670169), this [lines](https://github.com/kubernetes-sigs/external-dns/blob/master/endpoint/endpoint_test.go#L120-L174) on master are a wrong unit test, we can't have CNAME and A records with the same name.
> 
> @leonardocaylent to be clear the referenced lines may not be technically correct according to DNS rules, but for the specific test `TestFilterEndpointsByOwnerID` it doesn't matter because the test is for the "OwnerID" filter which is not responsible detecting invalid/duplicate names. The logic to remove duplicates should probably not be added to that function.

@cronik So I guess in that case changing the unit test to use only one type of records should be fine for this (same thing could be applied for avoid confusion on the L120-L174 in a different ticket) What do you think of the solution on endpoint.go? My concern is that the change impacts all providers and I'm not aware of use cases that needs duplicated endpoints

leonardocaylent commented:
Based on the feedback provided I was able to simplify the fix and set the unit tests according to our needs. Ready for re-review

cronik commented:
@leonardocaylent I don't think `FilterEndpointsByOwnerID` should be responsible for removing duplicates. It's only purpose is to filter by OwnerID as the function name indicates.  

I suggest creating a new function that performs the singular task of removing duplicates. If that is what the aws provider requires then apply that filter there. That's my 2 cents, I'll defer to the project maintainers though.

yurrriq commented:
Can we instead fix it `dedupSource.Endpoints()`? https://github.com/kubernetes-sigs/external-dns/blob/v0.14.1/source/dedupsource.go#L38

leonardocaylent commented:
@cronik @yurrriq Thanks for your comments. Taking another deep look I found the duplicates were on controller.go inside [Current](https://github.com/kubernetes-sigs/external-dns/blob/master/controller/controller.go#L219) & [Desired](https://github.com/kubernetes-sigs/external-dns/blob/master/controller/controller.go#L232). The new proposal is to add the function RemoveDuplicates on endpoint.go and use it on controller.go. I changed the unit tests to reflect the use cases on the new function

yurrriq commented on endpoint/endpoint.go:
```diff
@@ -343,3 +343,22 @@ type DNSEndpointList struct {
+
+// Apply filter based on EndpointKey (using DNSName, RecordType & SetIdentifier)
+func RemoveDuplicates(filtered []*Endpoint) []*Endpoint {
+	visited := make(map[EndpointKey]bool)
```
Is now the time to micro-optimize here? :joy:

yurrriq commented on controller/controller.go:
```diff
@@ -230,6 +231,7 @@ func (c *Controller) RunOnce(ctx context.Context) error {
 	verifiedARecords.Set(float64(vARecords))
 	verifiedAAAARecords.Set(float64(vAAAARecords))
 	endpoints, err = c.Registry.AdjustEndpoints(endpoints)
+	endpoints = endpoint.RemoveDuplicates(endpoints)
```
Should this happen before L226? Otherwise it seems the count might be incorrect, though I'm not sure what the consequence of that is.

```go
	sourceEndpointsTotal.Set(float64(len(endpoints)))
```

leonardocaylent replied:
Nice catch! I didn't see that, I need to check if AdjustEndpoints generates the duplicates or the duplicates are at L220. I have an easy way to see that, let me verify

yurrriq commented on endpoint/endpoint.go:
```diff
@@ -343,3 +343,22 @@ type DNSEndpointList struct {
+	visited := make(map[EndpointKey]struct{})
+	result := []*Endpoint{}
+
+	for _, ep := range filtered {
```
```suggestion
	for _, ep := range endpoints {
```

yurrriq commented on endpoint/endpoint.go:
```diff
@@ -343,3 +343,22 @@ type DNSEndpointList struct {
 }
+
+// Apply filter based on EndpointKey (using DNSName, RecordType & SetIdentifier)
+func RemoveDuplicates(filtered []*Endpoint) []*Endpoint {
```
Nitpick: I think `filtered` is a misleading name for this argument. `FilterEndpointsByOwnerID()` above uses the name `eps`, which I also dislike, but maybe it would be good to match.

My personal preference would be `endpoints`.

```suggestion
func RemoveDuplicates(endpoints []*Endpoint) []*Endpoint {
```

Also, I think this is a great opportunity to refactor `dedupSource.Endpoints()`, something like the following.

```go
// Endpoints collects endpoints from its wrapped source and returns them without duplicates.
func (ms *dedupSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error) {
	endpoints, err := ms.source.Endpoints(ctx)
	if err != nil {
		return nil, err
	}

	result := endpoint.RemoveDuplicates(endpoints)

	return result, nil
}
```

yurrriq commented on endpoint/endpoint.go:
```diff
@@ -343,3 +343,22 @@ type DNSEndpointList struct {
 	Items           []DNSEndpoint `json:"items"`
 }
+
+// Apply filter based on EndpointKey (using DNSName, RecordType & SetIdentifier)
```
Nitpick based on https://go.dev/doc/comment (though much of this project doesn't seem to follow this guidance anyway...): the doc comment should "start with [a] complete sentence naming the declared symbol" and "should not explain internal details".

Maybe something like the following.

```suggestion
// RemoveDuplicates returns a slice holding the unique endpoints.
```

yurrriq commented on endpoint/endpoint_test.go:
```diff
@@ -216,3 +216,136 @@ func TestIsOwnedBy(t *testing.T) {
+					bar2,
+				},
+			},
+			want: []*Endpoint{
```
So with this implementation, the first instance of an duplicate endpoint is the one that's kept. Is that an agreeable implementation detail? Thinking out loud here... `registry.Records()` is implemented by each provider, e.g. AWS, which calls (a variant of) `ListResourceRecordSets` under the hood which "sorts results first by DNS name with the labels reversed... When multiple records have the same DNS name, ListResourceRecordSets sorts results by the record type."

ref: https://pkg.go.dev/github.com/aws/aws-sdk-go/service/route53#hdr-Sort_order

If yes, then this PR seems a good solution to me.

leonardocaylent replied:
The concern is that even using the EndpointKey function, global and weighted records don't have enough test coverage and I'm not familiar if they use the same targets, but that's what dedupsource is using to generate the primary key, without the recordtype. The real questions here are "What use cases do we have that needs endpoints with the same "Key" but different targets?" "What should the endpoint's primary key be? Does it need to change across providers?"

leonardocaylent commented:
@yurrriq Thanks for your comments, after analyzing the plan.go you were correct on putting the light on that function changes. Here is the update: 

On version 0.13.6 and 0.14.0 always are duplicate endpoints on the variables 'records' at controller.go Line 206
The main difference is that changes.Delete on 0.13.6 was receiving duplicates, but the output was taking them out, apparently on this [function](https://github.com/kubernetes-sigs/external-dns/blob/v0.13.6/plan/plan.go#L157)
This [change](https://github.com/kubernetes-sigs/external-dns/blob/master/plan/plan.go#L196) seems like is not able to handle duplicates, so that's why a primary key on Endpoint is needed to correctly apply the filtering. 
In my opinion I think we need to add the string of targets as part of the primary key used on EndpointKey, but maybe it's not necessary.

mloiseleur commented:
@leonardocaylent Creating a new map of `Endpoint` may work and comes also with some impact on users with big clusters. I've mixed feelings about this, because it's a fix for an edge case, not the common case. 

A user should not have duplicated endpoint to start with, I try to detail why [here](https://github.com/kubernetes-sigs/external-dns/issues/4241#issuecomment-2012189562).

=> Wdyt about a fix that removes the root cause : the duplicated endpoint ?

leonardocaylent commented:
@mloiseleur I agree on that solution, about the aws.go provider, there is a part on `func suitableZones()` that has this comment 
`// Only select the best matching public zone`
Another ticket could be created to create and develop the feature for selecting the best matching zone (it doesn't care if it's public or private) so we only have the records created in the place we really need them. When users migate to the new version, external-dns should detect that this records are not needed and delete them on the plan.

leonardocaylent commented:
@yurrriq The root cause of the issue is that this [function](https://github.com/kubernetes-sigs/external-dns/blob/master/plan/plan.go#L167) lost the ability to remove the duplicate endpoints on the `current records` at 0.14.0
I tried adding the RemoveDuplicates function in [here](https://github.com/kubernetes-sigs/external-dns/blob/master/plan/plan.go#L330) but does break some unit tests:
```
--- FAIL: TestPlan (0.11s)
    --- FAIL: TestPlan/TestSyncSecondRound (0.00s)
        plan_test.go:951: expected ["foo 0 IN CNAME  v2 []"] to match []
    --- FAIL: TestPlan/TestSyncSecondRoundMigration (0.00s)
        plan_test.go:951: expected [] to match ["foo 0 IN CNAME  v1 []"]
```

szuecs commented:
Please follow up with https://github.com/kubernetes-sigs/external-dns/pull/4296#issuecomment-2020257118. Raffo said what he wants to see and if you want to get this done, then please follow his guideline.
We are happy to review the mentioned PR that you can link back to this one to make sure there is more priority from our side.

leonardocaylent commented:
@szuecs I've read @Raffo's comments. We were working with @yurrriq and @cronik  and we found the root cause. I'll put here a brief summary:
1)About the unit tests, they are a little missleading (having a same DNS name for an A record and a CNAME at the same time) but for the test purposes @cronik said the intention was to test the behavior, not the validation of being able to create both records. This [issue](https://github.com/kubernetes-sigs/external-dns/issues/4339) was created because is related to this but we think it's not a priority or needs to be addresed on the #4241
2)We found that duplicated endpoints were always at the records variable on controller.go, the real root cause of the issue was that on this Pull Request [multi-cluster-fix](https://github.com/kubernetes-sigs/external-dns/pull/3747) the plans started to be different:
```
		Behavior on 0.14.0 on Deletes
		------------------------------------------------------------------------------------------------------
		DNSName          | Current record                                      | Desired Records (Candidate) |
		------------------------------------------------------------------------------------------------------
		dnsname.zone.com |[dnsname.zone.com IN A]dnsname.zone.com IN A (dup ep)| []                          |
		------------------------------------------------------------------------------------------------------
	
		Behavior on 0.13.6 on Deletes
		------------------------------------------------------------------------------------------------------
		DNSName          | Current record                                      | Desired Records (Candidate) |
		------------------------------------------------------------------------------------------------------
		dnsname.zone.com | [dnsname.zone.com IN CNAME ]                        | []                          |
		------------------------------------------------------------------------------------------------------

```
3)I didn't see any new comments between @Raffo's comment and the new finding so I think that changes the solution. The things are on the current PR but the last commit (d9b7439) it's a Draft and it's really complex to explain on a comment. I'm still working on the solution because we find the real root cause only 5 days ago, which was merged into main 7 months ago. I'd really appreciate if some of the contributors with more experience have time to discuss in a huddle about this.
If that's not possible I'll post on the [original-issue](https://github.com/kubernetes-sigs/external-dns/issues/4241) to ask another review to this pull request, if that's not the way to contribute or I'm failing with some of the steps please reach out to me and I'll stop doing that. I'm new here and I just want contribute my grain of sand to k8s.

leonardocaylent commented:
@Raffo We could confirm that
1)The behavior on the new version is different:
Behavior on 0.13.6 on Deletes:

```
DNS Name : dnsname.zone.com
Current record: 
dnsname.zone.com 300 IN CNAME  k8s-ingre-1234abcd.us-west-2.elb.amazonaws.com [{aws/evaluate-target-health true} {alias true}] 
Desired Records (Candidate): []
```

Behavior on 0.14.0 on Deletes:

```
DNS Name : dnsname.zone.com
Current record: 
dnsname.zone.com 300 IN A  k8s-ingre-1234abcd.us-west-2.elb.amazonaws.com [{aws/evaluate-target-health true} {alias true}] 
dnsname.zone.com 300 IN A  k8s-ingre-1234abcd.us-west-2.elb.amazonaws.com [{aws/evaluate-target-health true} {alias true}] 
Desired Records (Candidate): []
```

2)There were 2 major changes on `plan.go`, the Record Type included in the plan, and the ability to remove duplicated endpoints on `changes.Delete`
3)The variable records on controller.go is working fine, it contains duplicated endpoints at that time because it's reflecting the current state of the DNS records on the provider, to have a low impact the change needs to be applied only on plan.go
4)This could be considered a fix to have the previous behavior prior to #3747 , but after having this in place, we could work on a feature for AWS "Creating the dns records for endpoints only in the Hosted Zone that has the best fit". A good starting point would be on the [suitableZones](https://github.com/kubernetes-sigs/external-dns/blob/master/provider/aws/aws.go#L1028) function, but I guess we should have some discussion about if having the records in only one hosted zone needs some type of regression. The ideal change would be if this new feature makes changes.Delete to detect the unnecessary records in the duplicated/triplicated hosted zones.

From my point of view, this PR is ready to review. Special thanks to all the persons that helped finding the root cause of this issue.

yurrriq reviewed (approved):
lgtm, I agree that any changes to the AWS provider to choose the most suitable zone are best addressed in a followup issue and PR

Raffo commented:
@leonardocaylent thank you for the update and the work with the rest of the community to figure out what the best path forward is. I followed the last updates and the change proposed seems reasonable to me. @szuecs @mloiseleur is the change fine for you as well? I feel that you followed this more closely, so I want to wait for your okay for this to proceed.

mloiseleur commented:
This change is working and lgtm. Even if it's an edge case, I understand it was working before.
I (really) hope that we will be able to fix the source and remove this code.

Raffo commented:
> I (really) hope that we will be able to fix the source and remove this code.

@mloiseleur yes, same. Do we want to get this in and work on the more permanent fix then? Does this require a release? Who's working on the second part? Sorry tons of questions 😅

leonardocaylent commented:
> > I (really) hope that we will be able to fix the source and remove this code.
> 
> @mloiseleur yes, same. Do we want to get this in and work on the more permanent fix then? Does this require a release? Who's working on the second part? Sorry tons of questions 😅

@Raffo I can work on the second part with help of someone does has experience and knowledge on the aws provider. The final fix will require the same amount of testing and debugging than this one (quite a lot)

leonardocaylent commented:
@Raffo @mloiseleur I agree that we want to get this merged in since the functionality of the plans changed and this will require a release. I also agree that we will be able to fix the source and remove this code entirely when we only generate one of each record in the correct hosted zone.

Raffo commented:
Ok so the plan is:

1. I take a final review at this.
2. If all is good, I merge.
3. I kick off the release process for v0.14.2.
4. We start working on the better fix and proceed accordingly. 

Does this sound correct?

leonardocaylent commented:
> Ok so the plan is:
> 
> 1. I take a final review at this.
> 2. If all is good, I merge.
> 3. I kick off the release process for v0.14.2.
> 4. We start working on the better fix and proceed accordingly.
> 
> Does this sound correct?

That sounds perfect, I created this [feature](https://github.com/kubernetes-sigs/external-dns/issues/4454) ticket to start the discussion about how to proceed with that. I explained with details the current behavior.

Raffo commented:
/lgtm
/approve
